src/utils/calculator.py

```diff
@@ -0,0 +1,131 @@
+from concurrent.futures import ThreadPoolExecutor, as_completed
+import numpy as np
+import os
+
+
+def load_agent_tasks(agent_name):
+    file_path = os.path.join(os.getcwd(), "agent_configs", f"{agent_name}_task.txt")
+    with open(file_path) as f:
+        task_inputs = f.readlines()
+        return task_inputs
+
+
+def get_numbers_concurrent(agent_list, agent_factory, agent_thread_pool):
+    agent_tasks = []
+    for agent_name, agent_num in agent_list:
+        task_input = load_agent_tasks(agent_name=agent_name)[0]
+        for i in range(agent_num):
+            agent_task = agent_thread_pool.submit(
+                agent_factory.run_agent,
+                agent_name,
+                task_input
+            )
+            agent_tasks.append(agent_task)
+    stats = {
+        'turnaround_times': [],
+        'waiting_times': [],
+        'request_waiting_times': [],
+        'request_turnaround_times': []
+    }
+
+    # Collect data
+    for result in as_completed(agent_tasks):
+        output = result.result()
+        stats['turnaround_times'].append(output["agent_execution_time"])
+        stats['waiting_times'].append(output["agent_waiting_time"])
+        stats['request_waiting_times'].extend(output["request_waiting_times"])
+        stats['request_turnaround_times'].extend(output["request_turnaround_times"])
+
+    # Compute averages and percentiles
+    def compute_metrics(data):
+        return {
+            'avg': np.mean(data),
+            'p90': np.percentile(data, 90),
+            'p99': np.percentile(data, 99)
+        }
+
+    metrics = {
+        'agent_turnaround_time': compute_metrics(stats['turnaround_times']),
+        'agent_waiting_time': compute_metrics(stats['waiting_times']),
+        'request_waiting_time': compute_metrics(stats['request_waiting_times']),
+        'request_turnaround_time': compute_metrics(stats['request_turnaround_times'])
+    }
+
+    return metrics
+
+
+def get_numbers_sequential(agent_list, agent_factory):
+    stats = {
+        'turnaround_times': [],
+        'waiting_times': [],
+        'request_waiting_times': [],
+        'request_turnaround_times': []
+    }
+
+    accumulated_time = 0
+    for agent_name, agent_num in agent_list:
+        task_input = load_agent_tasks(agent_name=agent_name)[0]  # Assuming first task relevant
+        for _ in range(agent_num):
+            output = agent_factory.run_agent(agent_name=agent_name, task_input=task_input)
+
+            agent_turnaround_time = output["agent_execution_time"] + accumulated_time
+            agent_waiting_time = output["agent_waiting_time"] + accumulated_time
+            rounds = output["rounds"]
+            fixed_offset = accumulated_time / rounds
+
+            # Adjust times by the accumulated time
+            request_waiting_times = [time + fixed_offset for time in output["request_waiting_times"]]
+            request_turnaround_times = [time + fixed_offset for time in output["request_turnaround_times"]]
+
```

Comment:
@dongyuanjushi could you help check the logic here?

Reply:
1. accumulated time should be: accumulated_time += (agent_turnaround_time - agent_waiting_time)
2.  request_waiting_times and request_turnaround_times should be that only the element at the first position in request level waiting times and turnaround times should add the accumulated_time and the remaining element should not be changed. As the agent will be executed sequential after the whole execution of its previous agent, so only the first request will be much longer, the remaining request waiting time and turnaround time should be the same?